File: Java/other/three_Sum_problem.java
/* 

3 Sum problem

For a given array nums[], your task is to find the triplets  
whose sum add up to the target value given. 
The output is the triplet of the numbers which add up to target. 
For example:- 
if A + B + C = target then the output will be A, B, C. 
It is given that all the sets of triplets should be unique. 

*/  

import java.util.Scanner; 
import java.util.Arrays; 
import java.util.ArrayList; 
import java.util.List; 

class three_Sum_problem 
{ 
    //method for printing triplet 
    public List<Integer[]> threeSum(int nums[], int target) 
    { 
        Set<List<Integer[]>> tripletEvaluated = new Hashset<>(); 

        //if the array has no elements then we simpliy return  
        //empty list 
        if(nums.length == 0) 
            return new ArrayList<>(tripletEvaluated); 

        //else 
        Arrays.sort(nums); 

        //using two pointer approach 
        for(int i=0; i<nums.length; i++) 
        { 
            int leftptr = i + 1; 

            int rightptr = nums.length - 1; 

            while(leftptr < rightptr)
            { 
                int complement = target - num[i]; 

                int sum = nums[leftptr] + nums[rightptr]; 

                if(sum == complement) 
                { 
                    tripletEvaluated.add(Arrays.asList(nums[i], nums[leftptr], nums[rightptr])); 
                    leftptr++; 
                    rightptr--; 
                } 

                else if(sum > complement) 
                { 
                    rightptr--; 
                } 

                //else if sum < complement 
                else 
                { 
                    leftptr++; 
                } 
            } 
        } 
        return new ArrayList<>(tripletEvaluated); 
    } 

    //driver method 
    public static void main(String []args) 
    { 

        Scanner sc = new Scanner(System.in); 

        System.out.println("Enter the size of the array: "); 
        int size = sc.nextInt(); 

        System.out.println("Enter the target value: "); 
        int target = sc.nextInt();

        int nums[] = new int[size]; 
        System.out.println("Enter the elements of array: "); 
        for(int i = 0; i < size; i++) 
        { 
            nums[i] = sc.nextInt(); 
        } 

        List<Integer[]> tripletEvaluated = threeSum(nums, target); 

        System.out.println("The triplet values are: "); 
        for(Integer[] triplets: tripletEvaluated) { 
        for(int numb: triplets) 
        { 
            System.out.print(numb + " "); 
        } 
            System.out.println(); 
        } 
    } 
} 

/* 

EXAMPLES:- 

Example 1: 
Input-- 
Enter the size of the array: 6 
Enter the target value: 0 
Enter the elements of array: 1 2 -1 0 -2 1 
Output-- 
The triplet values are: [1, -1, 0] 
                        [2, 0, -2] 
                        [1, 1, -2] 

Example 2: 
Input-- 
Enter the size of the array: 0 
Enter the target value: 0 
Enter the elements of array: 0 
Output-- 
The triplet values are: 0 

Example 3: 
Input-- 
Enter the size of the array: 5 
Enter the target value: 0 
Enter the elements of array: 1 -1 -1 0 1 
Output-- 
The triplet values are: [1, -1, 0] 

TIME COMPLEXITY --> O(N^2) 

*/ 

File: Python/search/three_sum_problem.py
"""
    Introduction:
        Two pointer technique is an optimization technique which is a really clever way of using
        brute-force to search for a particular pattern in a sorted list.

    Purpose:
        The code segment below solves the Three Sum Problem. We have to find a triplet out of the
        given list of numbers such that the sum of that triplet equals to another given value.
        The Naive Approach is to calculate the sum of all possible triplets and return the one with
        the required sum or None. This approach takes O(N^3) time to run. However, by using Two pointer
        search technique we can reduce the time complexity to O(N^2).

    Method:
        Three Sum Problem using Two Pointer Technique
"""

def three_sum_problem(numbers, sum_val):
    """
        Returns a triplet (x1,x2, x3) in list of numbers if found, whose sum equals sum_val, or None
    """

    # Sort the given list of numbers
    # Time complexity: O(N.logN)
    numbers.sort()
    size = len(numbers)

    # Two nested loops
    # Time complexity: O(N^2)
    for i in range(size-3+1):
        # Reduce the problem to two sum problem
        two_sum = sum_val - numbers[i]
        
        # Initialize the two pointers
        left = i+1
        right = size-1
        
        # Search in the array until the two pointers overlap
        while left < right:
            curr_sum = numbers[left] + numbers[right]

            # Update the pointers
            if curr_sum < two_sum:
                left += 1
            elif curr_sum > two_sum:
                right -= 1
            else:
                # Return the numbers that form the desired triplet
                return "({},{},{})".format(numbers[i], numbers[left], numbers[right])
    
    # No triplet found
    return None



def main():
    """
        Takes user input and calls the necessary function
    """

    # Take user input
    numbers = list(map(int, input("Enter the list of numbers: ").split()))
    sum_val = int(input("Enter the value of sum: "))

    # Find the triplet
    triplet = three_sum_problem(numbers, sum_val)

    if triplet is None:
        print(f"No triplet found with sum: {sum_val}")
    else:
        print(f"{triplet} is the triplet with sum: {sum_val}")


if __name__ == "__main__":
    # Driver code
    main()


"""
    Sample Input 1:
    Enter the list of numbers: 12 3 4 1 6 9
    Enter the value of sum: 24

    Sample Output 1:
    (3,9,12) is the triplet with sum: 24

    Time Complexity:
    For sorting:                O(N.log N)
    For two-pointer technique:  O(N^2)
    T = O(N.log N) + O(N^2)
    T = O(N^2)
"""